### PR TITLE
prometheus exporter convert instrumentation scope to otel_scope_info metric

### LINKED ISSUE #3273: Add Instrumentation Scope and Version as info metric and label in Prometheus exporter
Based on https://github.com/open-telemetry/opentelemetry-specification/pull/2703

> Prometheus exporters SHOULD generate an [Info](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#info)-typed
> metric named `otel_scope_info`. If present, Instrumentation Scope
> `name` and `version` MUST be added as `otel_scope_name` and
> `otel_scope_version` labels. Scope attributes MUST also be added as labels
> following the rules described in the [`Metric Attributes`](#metric-attributes)
> section below.
> 
> Prometheus exporters MUST add the scope name as the `otel_scope_name` label and
> the scope version as the `otel_scope_version` label on all metric points by
> default.
> 
> Prometheus exporters SHOULD provide a configuration option to disable the
> `otel_scope_info` metric and `otel_scope_` labels.

### PR DESCRIPTION
**Pull request by fatsheep9146**
fix #3273
related to https://github.com/open-telemetry/opentelemetry-go/issues/3161

Since #3285 already made an implementation for add target_info metric to prometheus exporter. I think I can wait for this pr merged and add otel_scope_info metric on the basic of his work.

### CODE REVIEW

**Review by dashpole — APPROVED** [review]
Note that this does not fully fix https://github.com/open-telemetry/opentelemetry-go/issues/3161.  I believe you also need to implement https://github.com/open-telemetry/opentelemetry-specification/pull/2890

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 ## [Unreleased]
 
+### Added
+
+- Add Instrumentation Scope and Version as info metric and label in Prometheus exporter. (#3273)
```
Comment (dashpole) [inline CHANGELOG.md:13]: Probably also good to mention the new option if people want to disable Instrumentation scope setting

**Comment by fatsheep9146** [comment]
> Note that this does not fully fix #3161. I believe you also need to implement [open-telemetry/opentelemetry-specification#2890](https://github.com/open-telemetry/opentelemetry-specification/pull/2890)

You are right, I mock a test as following.

If use two different meter to create two metric with same name and enable scope info, no errors happen
```
func TestMetricWithSameName(t *testing.T) {
	exporter, err := New()
	assert.NoError(t, err)

	provider := metric.NewMeterProvider(
		metric.WithReader(exporter),
	)

	httpCounter, err := provider.Meter("http").SyncInt64().Counter("error_count", instrument.WithUnit(unit.Dimensionless))
	assert.NoError(t, err)
	httpCounter.Add(context.TODO(), 1)

	sqlCounter, err := provider.Meter("sql").SyncInt64().Counter("error_count", instrument.WithUnit(unit.Dimensionless))
	assert.NoError(t, err)
	sqlCounter.Add(context.TODO(), 1)

	t.Logf("serving metrics at localhost:2223/metrics")
	http.Handle("/metrics", promhttp.Handler())
	err = http.ListenAndServe(":2223", nil)
	if err != nil {
		t.Fatalf("error serving http: %v", err)
		return
	}
}
```
curl metric port will produce following metric

```
# HELP error_count_ratio_total
# TYPE error_count_ratio_total counter
error_count_ratio_total{otel_scope_name="http",otel_scope_version=""} 1
error_count_ratio_total{otel_scope_name="sql",otel_scope_version=""} 1
```

But if I disable scope info like this 
```
func TestMetricWithSameName(t *testing.T) {
	exporter, err := New(WithoutScopeInfo())
	... // following codes keep unchanged
}
```
the error happned

```
An error has occurred while serving metrics:

collected metric "error_count_ratio_total" { counter:<value:1 > } was collected before with the same name and label values
```

I think I can solve this problem in another pr? Is that ok ? @dashpole @MrAlias

**Comment by dashpole** [comment]
> If use two different meter to create two metric with same name and enable scope info, no errors happen

Can you try with a different description on each?  I think that is why you aren't reproducing the issue

**Comment by fatsheep9146** [comment]
> Can you try with a different description on each? I think that is why you aren't reproducing the issue.

For two `Counter` of two different `meter` with different `Description`, an error happend 

code
```go
        ...
	httpCounter, err := provider.Meter("http").SyncInt64().Counter("error_count", instrument.WithUnit(unit.Dimensionless), instrument.WithDescription("http error count"))
	httpCounter.Add(context.TODO(), 1)

	sqlCounter, err := provider.Meter("sql").SyncInt64().Counter("error_count", instrument.WithUnit(unit.Dimensionless), instrument.WithDescription("sql error count"))
	sqlCounter.Add(context.TODO(), 1)
        ...
```
error 
```
collected metric error_count_ratio_total label:<name:"otel_scope_name" value:"sql" > label:<name:"otel_scope_version" value:"" > counter:<value:1 >  has help "sql error count" but should have "http error count"
```

**Comment by fatsheep9146** [comment]
> Can you also try with different Counter vs UpDownCounter?

For one `Counter` and another `UpDownCounter` of two different meter with same `name`, same `description`, no error happend

code
```go

httpCounter, err := provider.Meter("http").
  SyncInt64().Counter(
  "error_count",
  instrument.WithUnit(unit.Dimensionless))
httpCounter.Add(context.TODO(), 1)

sqlCounter, err := provider.Meter("sql").
  SyncInt64().UpDownCounter(
  "error_count",
  instrument.WithUnit(unit.Dimensionless))
sqlCounter.Add(context.TODO(), 1)
```

result
```
# HELP error_count_ratio
# TYPE error_count_ratio gauge
error_count_ratio{otel_scope_name="sql",otel_scope_version=""} 1
# HELP error_count_ratio_total
# TYPE error_count_ratio_total counter
error_count_ratio_total{otel_scope_name="http",otel_scope_version=""} 1
...
```

File: exporters/prometheus/exporter.go [hunk 2]
```diff
@@ -119,27 +131,48 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
...
+				scopeInfo = c.scopeInfos[scopeMetrics.Scope]
+			} else {
+				scopeInfo, err = createScopeInfoMetric(scopeMetrics.Scope)
+				if err != nil {
+					otel.Handle(err)
+				}
+				c.scopeInfos[scopeMetrics.Scope] = scopeInfo
+			}
+			ch <- scopeInfo
+			keys = append(keys, scopeInfoKeys...)
```
Comment (MrAlias) [inline exporters/prometheus/exporter.go:148]:
```suggestion
			keys = scopeInfoKeys
```

File: exporters/prometheus/exporter.go [hunk 3]
```diff
@@ -119,27 +131,48 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
...
+			values = append(values, getScopeKeyValues(scopeMetrics.Scope)...)
```
Comment (MrAlias) [inline exporters/prometheus/exporter.go:149]:
```suggestion
			values = [2]string{scopeMetrics.Scope.Name, scopeMetrics.Scope.Version}
```

File: exporters/prometheus/exporter.go [hunk 4]
```diff
@@ -119,27 +131,48 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
...
 			case metricdata.Gauge[float64]:
-				addGaugeMetric(ch, v, m, c.getName(m))
+				addGaugeMetric(ch, v, m, keys, values, c.getName(m))
 			}
 		}
 	}
 }
 
-func addHistogramMetric(ch chan<- prometheus.Metric, histogram metricdata.Histogram, m metricdata.Metrics, name string) {
+func addHistogramMetric(ch chan<- prometheus.Metric, histogram metricdata.Histogram, m metricdata.Metrics, ks, vs []string, name string) {
```
Comment (MrAlias) [inline exporters/prometheus/exporter.go:169]:
```suggestion
func addHistogramMetric(ch chan<- prometheus.Metric, histogram metricdata.Histogram, m metricdata.Metrics, ks, vs [2]string, name string) {
```

etc.

File: exporters/prometheus/exporter_test.go [hunk 5]
```diff
@@ -262,7 +300,10 @@ func TestPrometheusExporter(t *testing.T) {
 				metric.WithResource(res),
 				metric.WithReader(exporter, customBucketsView, defaultView),
 			)
-			meter := provider.Meter("testmeter")
+			meter := provider.Meter(
+				"testmeter",
+				otelmetric.WithInstrumentationVersion("v0.1.0"),
+			)
```
Comment (MrAlias) [inline exporters/prometheus/exporter_test.go:306]: Can you add another test that validates multiple scopes from different meters are assigned correctly to the instruments they create?

**Comment by dashpole** [comment]
> what's the expected behavior here? Just drop the second attempt to collect the second instrument and log an error?

From https://github.com/open-telemetry/opentelemetry-specification/pull/2890

> Example approach
> * Keep a map of metric name -> Description when constructing a batch of metrics for a scrape.
> * Each time a new metric (with Description) is added, look up the name in the map:
>     * If it isn't found, add it to the map
>     * If it is found, and the desc is the same, do nothing
>     * If it is found, and has conflicting type information, drop the metric + log a warning
>     * If it is found, and has conflicting help or unit, use the stored help + unit instead of the provided one + log a warning.

Basically, we need the description of all metrics with the same name to match.

File: exporters/prometheus/exporter_test.go [hunk 6]
```diff
@@ -306,3 +347,35 @@ func TestSantitizeName(t *testing.T) {
...
+// 	// sqlCounter.Add(context.TODO(), 1)
+
+// 	t.Logf("serving metrics at localhost:2223/metrics")
+// 	http.Handle("/metrics", promhttp.Handler())
+// 	err = http.ListenAndServe(":2223", nil)
+// 	if err != nil {
+// 		t.Fatalf("error serving http: %v", err)
+// 		return
+// 	}
+// }
```
Comment (MrAlias) [inline exporters/prometheus/exporter_test.go:381]:
```suggestion
```

File: exporters/prometheus/exporter_test.go (hunk 5 above)
Comment (MrAlias) [reply exporters/prometheus/exporter_test.go:307]: This looks unresolved. @fatsheep9146 are you still planning to address this or did I miss the addition that already did?

**Comment by MrAlias** [comment]
> > what's the expected behavior here? Just drop the second attempt to collect the second instrument and log an error?
> 
> From [open-telemetry/opentelemetry-specification#2890](https://github.com/open-telemetry/opentelemetry-specification/pull/2890)
> 
> > Example approach
> > 
> > * Keep a map of metric name -> Description when constructing a batch of metrics for a scrape.
> > * Each time a new metric (with Description) is added, look up the name in the map:
> >   
> >   * If it isn't found, add it to the map
> >   * If it is found, and the desc is the same, do nothing
> >   * If it is found, and has conflicting type information, drop the metric + log a warning
> >   * If it is found, and has conflicting help or unit, use the stored help + unit instead of the provided one + log a warning.
> 
> Basically, we need the description of all metrics with the same name to match.

:+1: 

@dashpole are you good with merging this and then adding that tracking in another PR? That seems to make sense to me.

File: exporters/prometheus/exporter_test.go (hunk 5 above)
Comment (fatsheep9146) [reply exporters/prometheus/exporter_test.go:308]: The test is added, `TestMultiScopes`.

**Comment by dashpole** [comment]
> @dashpole are you good with merging this and then adding that tracking in another PR? That seems to make sense to me.

Yes.

**Review by MrAlias — APPROVED** [review]

**Pull request merged**